### DISCUSSION TITLE
About Animation TimeLine Script

### DISCUSSION THREAD

**Discussion by shenshengyi** [iTwin/itwinjs-core#3006] [Q&A] [answered]
I use the Schedule TimeLine Script feature in iTwin.js version 2.19.26 and it works fine.
Then I ported the same frontend code to iTwin.js version ^3.0.0-dev.17X , the animation won't run and the program doesn't report any errors.
I'm wondering what could be the reason for this failure? Thanks.

My frontend code: 
export async function HandleSelectedScript(scriptID: string) {
  const vp = IModelApp.viewManager.selectedView;
  if (vp) {
    vp.displayStyle.settings.renderTimeline = scriptID;
    await vp.displayStyle.changeRenderTimeline(undefined);
  }
}
itwin.js 2.19.26 The correct video is as follows: 
https://user-images.githubusercontent.com/26344907/148648691-871c3c45-397f-4cbc-9f14-a92a13442d02.mp4

**Comment by pmconne** [comment]
> `await vp.displayStyle.changeRenderTimeline(undefined);`

Passing `undefined` means you don't want any animation script applied to the view. Try this:
```
export async function HandleSelectedScript(scriptID: string) {
  const vp = IModelApp.viewManager.selectedView;
  if (vp)
    await vp.displayStyle.changeRenderTimelineId(scriptID);
}
```

**Reply by shenshengyi** [reply]
> > `await vp.displayStyle.changeRenderTimeline(undefined);`
> 
> Passing `undefined` means you don't want any animation script applied to the view. Try this:
> 
> ```
> export async function HandleSelectedScript(scriptID: string) {
>   const vp = IModelApp.viewManager.selectedView;
>   if (vp)
>     await vp.displayStyle.changeRenderTimelineId(scriptID);
> }
> ```

hi, after this modification, iTwin.js version ^3.0.0-dev.17X still cannot work. And iTwin 2.19.26 doesn't work anymore.

**Comment by pmconne** [comment]
I am going to need a more precise definition of "work" and "doesn't work", and any other details (e.g., is this code running in iTwin Viewer? Are you relying on the appui timeline component? Does it occur if the view already has a script and you changed it to a different one, or only if the view doesn't have a script? etc).
I cannot repro in display-test-app.

## Master

Pull master. Apply the following diff:
```
diff --git a/test-apps/display-test-app/public/locales/en/SVTTools.json b/test-apps/display-test-app/public/locales/en/SVTTools.json
index 635b6691cc..7d9765b5a8 100644
--- a/test-apps/display-test-app/public/locales/en/SVTTools.json
+++ b/test-apps/display-test-app/public/locales/en/SVTTools.json
@@ -73,6 +73,9 @@
     "ShutDown": {
       "keyin": "dta shutdown"
     },
+    "DtaDebug": {
+      "keyin": "dta debug"
+    },
     "Exit": {
       "keyin": "dta exit"
     },
@@ -152,4 +155,4 @@
       "keyin": "dta macro"
     }
   }
-}
\ No newline at end of file
+}
diff --git a/test-apps/display-test-app/src/frontend/App.ts b/test-apps/display-test-app/src/frontend/App.ts
index c12f9e0f19..1a9de61a04 100644
--- a/test-apps/display-test-app/src/frontend/App.ts
+++ b/test-apps/display-test-app/src/frontend/App.ts
@@ -191,6 +191,28 @@ class ShutDownTool extends Tool {
   }
 }
 
+class DebugTool extends Tool {
+  public static override toolId = "DtaDebug";
+  public static override get maxArgs() { return undefined; }
+
+  public override async run(args: any[]) {
+    const vp = IModelApp.viewManager.selectedView;
+    if (vp) {
+      const was = vp.displayStyle.settings.renderTimeline;
+      vp.displayStyle.settings.renderTimeline = args[0];
+      const then = vp.displayStyle.settings.renderTimeline;
+      await vp.displayStyle.changeRenderTimeline(undefined);
+      console.log(`${was} ${then} ${vp.displayStyle.settings.renderTimeline}`);
+    }
+
+    return true;
+  }
+
+  public override async parseAndRun(...args: string[]) {
+    return this.run(args);
+  }
+}
+
 class ExitTool extends Tool {
   public static override toolId = "Exit";
 
@@ -278,6 +300,7 @@ export class DisplayTestApp {
       CloseIModelTool,
       CloseWindowTool,
       CreateWindowTool,
+      DebugTool,
       DockWindowTool,
       DrawingAidTestTool,
       EditingScopeTool,
```

Build and run display-test-app. Open your iModel. Fit the view.
Key in `dta debug 0x299` to set the schedule script. You'll notice the view flashes briefly as it reacts to the change by loading new tiles.
Open the animation panel and click the "play" button. Observe your balloon take flight.
![image](https://user-images.githubusercontent.com/22944042/148689667-370dad76-a369-43e9-8580-75fc3aa9d6ba.png)

While animation is playing, key in `dta debug 0x29a`. Observe your balloon returns to the ground and your trucks and bus begin moving.

## 2.19.27

Now pull release/2.19.x (aka 2.19.27) and apply the following nearly identical diff:
```
diff --git a/test-apps/display-test-app/public/locales/en/SVTTools.json b/test-apps/display-test-app/public/locales/en/SVTTools.json
index 92bdcec5dc..a03df2ebf3 100644
--- a/test-apps/display-test-app/public/locales/en/SVTTools.json
+++ b/test-apps/display-test-app/public/locales/en/SVTTools.json
@@ -70,6 +70,9 @@
     "ShutDown": {
       "keyin": "dta shutdown"
     },
+    "DtaDebug": {
+      "keyin": "dta debug"
+    },
     "ToggleShadowMapTiles": {
       "keyin": "dta shadow tiles"
     },
@@ -125,4 +128,4 @@
       "keyin": "dta viewport sync"
     }
   }
-}
\ No newline at end of file
+}
diff --git a/test-apps/display-test-app/src/frontend/App.ts b/test-apps/display-test-app/src/frontend/App.ts
index 1e67593c74..e2ec0ac70b 100644
--- a/test-apps/display-test-app/src/frontend/App.ts
+++ b/test-apps/display-test-app/src/frontend/App.ts
@@ -137,6 +137,28 @@ class ShutDownTool extends Tool {
   }
 }
 
+class DebugTool extends Tool {
+  public static override toolId = "DtaDebug";
+  public static override get maxArgs() { return undefined; }
+
+  public override run(args: any[]): boolean {
+    const vp = IModelApp.viewManager.selectedView;
+    if (vp) {
+      const was = vp.displayStyle.settings.renderTimeline;
+      vp.displayStyle.settings.renderTimeline = args[0];
+      const then = vp.displayStyle.settings.renderTimeline;
+      vp.displayStyle.changeRenderTimeline(undefined);
+      console.log(`${was} ${then} ${vp.displayStyle.settings.renderTimeline}`);
+    }
+
+    return true;
+  }
+
+  public override parseAndRun(...args: string[]) {
+    return this.run(args);
+  }
+}
+
 export class DisplayTestApp {
   public static tileAdminProps: TileAdmin.Props = {
     retryInterval: 50,
@@ -210,6 +232,7 @@ export class DisplayTestApp {
       CloseIModelTool,
       CloseWindowTool,
       CreateWindowTool,
+      DebugTool,
       DockWindowTool,
       DrawingAidTestTool,
       EditingScopeTool,
```

Repeat all of the previous steps and make the same observations. (You will see an error about browser authorization - it can be ignored).

## Note

It absolutely is a bug that changeRenderTimeline(undefined) does not in fact set the timeline Id to undefined. But that appears to be entirely unrelated to your problem and exists on both master and 2.19.27.

**Comment by pmconne** [comment, marked as answer]
@shenshengyi #3030 will address the **Note** in my previous reply and render your code (which should never have worked) non-working.
Please adjust your code to:
```
export async function HandleSelectedScript(scriptID: string) {
  const vp = IModelApp.viewManager.selectedView;
  if (vp)
    await vp.displayStyle.changeRenderTimeline(scriptID);
}
```

If problems persist please provide the additional information previously requested.

**Reply by shenshengyi** [reply]
@pmconne  hi，When I say not working, I mean that this timeline panel will pop up automatically after the function call is triggered in itwin2.19.26.
The same code, in iTwin.js version ^3.0.0-dev.17X , but the timeline panel doesn't appear, so I can't drag the timeline to control the animation. 

I updated the dependency package, but I didn't find the function changeRenderTimelineId, and the code still doesn't work properly.

**Reply by pmconne** [reply]
The function is called changeRenderTimeline, not changeRenderTimelineId. My change is in 3.0.0-dev.184.
@bsteinbk or @NancyMcCallB made some changes in 3.0 to how the timeline component behaves. They should be able to help you, if you give them a way to reproduce your problem.

**Reply by shenshengyi** [reply]
OK,Thanks.

**Reply by bsteinbk** [reply]
@shenshengyi - we had multiple request only to show the overlays if the stage opted in to showing them, so that is what has changed n 3.0. Now when the ContentProps are defined for a stage users must opt in to showing each supported overlay type. In the example below all default view overlays are enabled.

```
      const thisContentProps: ContentProps = {
        id: "imodel-view-default",
        classId: IModelViewportControl,
        applicationData:
        {
          viewState, iModelConnection,
          featureOptions:
          {
            defaultViewOverlay: {
              enableScheduleAnimationViewOverlay: true,
              enableAnalysisTimelineViewOverlay: true,
              enableSolarTimelineViewOverlay: true,
            },
          },
        },
      };
```

**Reply by shenshengyi** [reply]
Thanks a lot, it's resolved.
